src/hotspot/share/memory/metaspaceShared.cpp

```diff
@@ -776,6 +777,132 @@ bool MetaspaceShared::try_link_class(InstanceKlass* ik, TRAPS) {
   }
 }
 
+void append_strings(char* buffer, size_t buff_len,  const char* arg) {
+  char* start = buffer + strlen(buffer);
+  snprintf(start, buff_len, "%s ", arg);
+}
```

Comment:
Maybe move the above function to the StringUtils class under share/utilities?
Use `os::snprintf()` instead of `snprintf()`?

Reply:
The calculation is also wrong, this would overflow. You need:
```
  char* start = buffer + strlen(buffer);
  snprintf(start, buff_len - (start - buffer), "%s ", arg);
```
- and maybe add an assert that strlen(buf) < bufflen. 
- and as Ioi wrote, I'd use either one of os::snprintf or jio_snprintf since both guarantee zero termination on truncation.
- or, just use strncat()

Reply:
The new solution is using CDS.java to do the work.